### ISSUE TITLE
deprecate unused twisted.spread.jelly._unjelly_instance

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7653]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7653|
|Type|enhancement|
|Created|2014-09-27 13:08:11Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-jelly-instance-support-7653|

It is unused since 2003 and will not be ported to Python 3

See https://twistedmatrix.com/pipermail/twisted-python/2003-March/003345.html
and 

Attachments:

* [7653a.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d32/d32765fda7881b73559d0618e18e570f757a9a08/84b4ec889a2c8b072dfc2f91485902707915899d.patch) (2569 bytes) - added by wolfgang61 on 2014-09-27 13:10:45Z - 
* [7653b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d32/d32765fda7881b73559d0618e18e570f757a9a08/d0a2f1cfa04b6d2cf83e716c0c026449155a7a86.patch) (2590 bytes) - added by wolfgang61 on 2014-09-28 10:36:50Z - first patch was wrong
* [7653c.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d32/d32765fda7881b73559d0618e18e570f757a9a08/397ae318723b0456f2c0df70276809191df2797f.patch) (2522 bytes) - added by wolfgang61 on 2014-09-29 09:52:03Z - 
* [7653c.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d32/d32765fda7881b73559d0618e18e570f757a9a08/a69ce200b90dfe4e0c6b82cb6e0cc1702e76c9ad.patch) (2807 bytes) - added by wolfgang61 on 2014-09-29 10:00:46Z - 
* [7653d.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d32/d32765fda7881b73559d0618e18e570f757a9a08/a69b9bc99c59560281d62cfeab15a3a0c3cfc92d.patch) (3750 bytes) - added by wolfgang61 on 2014-09-29 22:14:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7653 7653
type__enhancement enhancement
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__None None
branch__branches_deprecate_jelly_instance_support_7653 branches/deprecate-jelly-instance-support-7653
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__spread spread
time__1411823291308289 1411823291308289
changetime__1445217216921098 1445217216921098
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Thanks.  I'm removing the Twisted 14.1.0 milestone because milestones are used to mark release blockers.  This ticket is normal development and there's no reason 14.1.0 _must_ include a resolution for this.  I'm also changing the ticket type because the existence of old functionality which no one is interested in anymore doesn't seem like a _defect_ to me (unless I've missed something, I don't think you can point to any particular behavior of Twisted and say "That is wrong!  That is because `_jelly_instance`!).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

  1. Please add a unit test for the deprecation warning.
  2. I'm not sure it makes sense to add the Python 3 check here.  The jelly test suite isn't being run on Python 3 yet.  Ideally, the test suite would be running to demonstrate that simply not defining this method causes some behavior to change from incorrect to correct.  As is, the conditional definition introduced by this patch is effectively untested code - which is not allowed by the Twisted coding standard or development process.

Also, in general, private APIs in Twisted do not need to be deprecated before removal.  That's a large part of the point of having separate private and public APIs.  Any application code directly calling `_unjelly_instance` is taking the risk that Twisted will break it with no warning.

But in this case, the private API is part of the implementation of a public protocol.  What's really being deprecated here is support for the `instance` atom in jelly.  The jelly protocol is public (because `twisted.spread.jelly.jelly` and `twisted.spread.jelly.unjelly` are public and have certain behaviors defined by the protocol).  I think that in the deprecation warning, rather than talking about `_unjelly_instance` it would make more sense to talk about "unjellying `instance` atoms".  And in the news fragment, it would make more sense to talk about `twisted.spread.jelly` or `Twisted Spread` or `The Jelly implementation`.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|wolfgang61 set owner to @exarkun|
|-|-|

I hope this is better. Ideally the test should call jelly.unjelly but then I would need a correctly encoded instance atom and since no code exists for generating that, this could be a bit difficult.

The jelly test suite actually runs fine with Python 3, also banana and pb, but only locally on my system. I also have a real app running with Python3 for server and Python2 for client.

But the _PY3 specific code is now removed from this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

You can exercise the `instance` code-path like this:

```
>>> unjelly(["instance", ["class", "something.Foo"], ["dictionary"]])
<something.Foo instance at 0x2421170>
>>> 
```

> The jelly test suite actually runs fine with Python 3, also banana and pb, but only locally on my system.

Right.  It needs to run on BuildBot in order for changes that are exercised by it to be checked in to version control. :)

Other comments:

  1. Twisted has switched away from the "I ..." style of docstring.
  1. There's no meaningful stacklevel for this deprecation warning.  stack levels are for deprecations of Python APIs.  This is a deprecation of a protocol feature.  Take a look at `twisted/internet/_sslverify.py`, around line 170, for an example of a warning where no stack frame makes sense.  Please also open a ticket for documenting the trick used there.
  1. The indentation of the multi-line `warnings.warn` call doesn't follow the Twisted coding standard.
  1. If you want to deprecate `twisted.spread.jelly.instance_atom` too, then it needs a warning too.  See `twisted.python.deprecate.deprecatedModuleAttribute`.
  1. The news fragment doesn't follow the style guidelines.  Try "twisted.spread.jelly's support for unjellying "instance" atoms is now deprecated."

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|wolfgang61 set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_4):
>   1. There's no meaningful stacklevel for this deprecation warning.  stack levels are for deprecations of Python APIs.  This is a deprecation of a protocol feature.  Take a look at `twisted/internet/_sslverify.py`, around line 170, for an example of a warning where no stack frame makes sense.  Please also open a ticket for documenting the trick used there.

Not sure how to do that. I removed stacklevel. Why should I need that trick? There is no exception involved anymore.

>   1. The indentation of the multi-line `warnings.warn` call doesn't follow the Twisted coding standard.

Can you please point me to that? I searched [https://twistedmatrix.com/documents/14.0.0/core/development/policy/coding-standard.html] for things like indent, long string, multi but did not find anything. twistedchecker says this is OK.

The rest is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [43238]) Branching to 'deprecate-jelly-instance-support-7653'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [43239]) Apply 7653d.patch 

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> Not sure how to do that. I removed stacklevel. Why should I need that trick? There is no exception involved anymore.

The stack level (which just defaults to 1) determines which source code line is displayed with the warning.  The behavior you get with the default stack level is this:

```
>>> unjelly(["instance", ["class", "__main__.Foo"], ["dictionary"]])
twisted/spread/jelly.py:886: DeprecationWarning: Unjellying an instance is deprecated since Twisted 14.1.0 and unsupported for Python 3
  DeprecationWarning)
```

The idiom is that the source line displayed is the line that is invoking the deprecated behavior.  It is technically correct that twisted/spread/jelly.py:886 is invoking the deprecated behavior - but only at the request of some other code.  It doesn't help an application developer track down which code of their own they need to fix to see this line.  Compare this to a simpler case, where a library `foo` has deprecated a function `bar` which is being used by an application:

```
import foo

def myappcode():
    foo.bar()
```

If `foo.bar` is deprecated with a warning with a stack level of 1, then the line blamed for this warning is the `    foo.bar()` line.  That makes a lot of sense.  It's exactly what the application developer needs to fix.

In the case of jelly, if an application developer ever sees this warning then what they likely need to fix is the Jelly implementation on the other end of a network connection.  There's no useful local source code line that can be displayed.  The trick is to prevent any line from being displayed at all in the hopes that this will be less confusing to whoever sees it and wants to correct the issue.

> Can you please point me to that? 

The Twisted coding standard defers to PEP 8 on issues it doesn't cover.  The relevant PEP 8 link is <http://legacy.python.org/dev/peps/pep-0008/#indentation>.

> twistedchecker says this is OK.

That makes me sad since this kind of mundane near-irrelevancy is the reason twistedchecker exists.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [43240]) since the FQPN already appears in the message this generates, just go with the name of the thing being deprecated in the namespace it is being deprecated in

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [43241]) remove some unimportant or incorrect information from the _unjelly_instance method docstring; reformat the warnings.warn call

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [43242]) Remove extra info from new test method docstrings.  Reformat long lines.

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [43243]) Flush warnings from this method only to ensure stack level is right.  Also fix the message that I broke earlier by changing the implementation.

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [43244]) Switch to warnings.warn_explicit to avoid blaming the wrong source line for the problem

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [43245]) Talk about the "instance atom" instead of "an instance" to minimize heart attacks.  Suggest a course of action too.


Don't talk about Python 3 at all.  That can come later, or hopefully may even
have been deleted by that point.

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [43246]) Grammar.  Also fix the test.

refs [#7653](https://github.com/twisted/twisted/issues/7653)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Thanks.  This looks good to me now.  Merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [43253]) Merge deprecate-jelly-instance-support-7653

Author: wolfgang61, exarkun
Reviewer: exarkun
Fixes: [#7653](https://github.com/twisted/twisted/issues/7653)

Formally deprecate the `instance` atom in the Jelly protocol.
Twisted has not emitted this token for at least ten years.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
